notify/sns/sns.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+	commoncfg "github.com/prometheus/common/config"
+)
+
+// Notifier implements a Notifier for SNS notifications.
+type Notifier struct {
+	conf    *config.SNSConfig
+	tmpl    *template.Template
+	logger  log.Logger
+	client  *http.Client
+	retrier *notify.Retrier
+	isFifo  *bool
+}
+
+// New returns a new SNS notification handler.
+func New(c *config.SNSConfig, t *template.Template, l log.Logger, httpOpts ...commoncfg.HTTPClientOption) (*Notifier, error) {
+	client, err := commoncfg.NewClientFromConfig(*c.HTTPConfig, "sns", append(httpOpts, commoncfg.WithHTTP2Disabled())...)
+	if err != nil {
+		return nil, err
+	}
+	return &Notifier{
+		conf:    c,
+		tmpl:    t,
+		logger:  l,
+		client:  client,
+		retrier: &notify.Retrier{},
+	}, nil
+}
+
+func (n *Notifier) Notify(ctx context.Context, alert ...*types.Alert) (bool, error) {
+	var (
+		err   error
+		data                           = notify.GetTemplateData(ctx, n.tmpl, alert, n.logger)
+		tmpl                           = notify.TmplText(n.tmpl, data, &err)
+		creds *credentials.Credentials = nil
+	)
+	if n.conf.Sigv4.AccessKey != "" && n.conf.Sigv4.SecretKey != "" {
+		creds = credentials.NewStaticCredentials(n.conf.Sigv4.AccessKey, string(n.conf.Sigv4.SecretKey), "")
+	}
+
+	attributes := make(map[string]*sns.MessageAttributeValue, len(n.conf.Attributes))
+	if len(n.conf.Attributes) > 0 {
+		for k, v := range n.conf.Attributes {
+			attributes[tmpl(k)] = &sns.MessageAttributeValue{DataType: aws.String("String"), StringValue: aws.String(tmpl(v))}
+		}
+
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Region:      aws.String(tmpl(n.conf.Sigv4.Region)),
+			Credentials: creds,
+			Endpoint:    aws.String(tmpl(n.conf.APIUrl)),
+		},
+		Profile: n.conf.Sigv4.Profile,
+	})
+
+	if _, err := sess.Config.Credentials.Get(); err != nil {
+		return false, fmt.Errorf("could not get SigV4 credentials: %w", err)
+	}
+
+	if n.conf.Sigv4.RoleARN != "" {
+		sess.Config.Credentials = stscreds.NewCredentials(sess, n.conf.Sigv4.RoleARN)
+	}
+
+	// Max message size for a message in a SNS publish request is 256KB, except for SMS messages where the limit is 1600 characters/runes.
+	messageSize := 256 * 1024
+	client := sns.New(sess, &aws.Config{Credentials: creds})
+	publishInput := &sns.PublishInput{}
+
+	if n.conf.TopicARN != "" {
+		publishInput.SetTopicArn(tmpl(n.conf.TopicARN))
+
+		if n.isFifo == nil {
+			checkFifo, err := checkTopicFifoAttribute(client, n.conf.TopicARN)
```

Comment:
You have to evaluate the template here as well. You may do it once and then use the compiled one both for `SetTopicArn()` and `checkTopicFifoAttribute()`.

```suggestion
			checkFifo, err := checkTopicFifoAttribute(client, tmpl(n.conf.TopicARN))
```